runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java

```diff
@@ -0,0 +1,300 @@
+        /**
+         * This ID represents the entry and its order in the Queue.
+         * This implies that it is unique and comparable with other IDs
+         * returned from CorfuQueue methods with respect to its enqueue order.
+         * However it cannot be compared with ID returned from the enqueue method for ordering
+         * because if this method is wrapped in a transaction, the order is established only later.
+         */
+        @Getter
+        final UUID ID;
+
+        @Getter
+        final E entry;
+
+        public String toString() {
+            return String.format(ID.toString()+"=>"+entry.toString());
+        }
+
+        CorfuQueueRecord(UUID id, E entry) {
+            this.ID = id;
+            this.entry = entry;
+        }
+    }
+
+    final private static int MAX_BITS_FOR_INDEX = 24;
+    /**
+     * Returns a List of CorfuQueueRecords sorted by the order in which the enqueue materialized.
+     *
+     * To re-constitute the commit order in the presence of transactions the CorfuRecords returned
+     * here have their UUID fields composed of two parts:
+     *     +----------------------------------------------------------------------------+
+     *     | Stream snapshot |Index in snapshot |     ID of the entry in the map        |
+     *     +----------------------------------------------------------------------------+
+     *     <----5 bytes------><----3 bytes-----><------- 8 bytes ----------------------->
+     *
+     * <p>This function currently does not return a view like the java.util implementation,
+     * and changes to the entryList will *not* be reflected in the map. </p>
+     *
+     * @param maxEntries - Limit the number of entries returned from start of the queue
+     *                   -1 implies all entries
+     * @return List of Entries sorted by their enqueue order
+     */
+    @Accessor
+    public List<CorfuQueueRecord<E>> entryList(int maxEntries) {
+        if (maxEntries == -1) {
+            maxEntries = Integer.MAX_VALUE;
+        }
+
+        long snapshotVersion = 0;
+        if (TransactionalContext.getCurrentContext() != null) {
+            snapshotVersion = TransactionalContext.getCurrentContext().getSnapshotTimestamp().getSequence();
+        }
+        List<CorfuQueueRecord<E>> copy = new ArrayList<>(Math.min(mainMap.size(), maxEntries));
+        int index = 0;
+        for (Map.Entry<Long, E> entry : mainMap.entrySet()) {
+            if (++index >= maxEntries) {
+                break;
+            }
+            UUID orderedId = new UUID((snapshotVersion<<MAX_BITS_FOR_INDEX) | index,
+            entry.getKey());
+            CorfuQueueRecord<E> record = new CorfuQueueRecord<>(orderedId, entry.getValue());
+            copy.add(record);
+        }
+        return copy;
+    }
+
+    @Accessor
+    public boolean isEmpty() {
+        return mainMap.isEmpty();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Accessor
+    public boolean containsKey(UUID key) {
```

Comment:
Need the `@ConflictParameter`

Reply:
How do I specify a @ConflictParameter on just the least significant bytes of an UUID?
That's why its currently separated into a public method calling an internal method with an @Accessor.

Reply:
BTW, you have to double check that `@Accessor` works correctly in this case.  There is a limitation (at least in hibernate/spring I have seen it), when you try to execute an annotated method you expect the logic related to the annotation will be triggered, which not happens, unfortunately (there is a limitation in Java annotation processing)

Reply:
Ok @xnull let me re-write the CorfuQueue completely as per your suggestion :)